### PR TITLE
native bindings - interfaces

### PR DESCRIPTION
**Pull request by akiezun**
@droazen please review

### CODE REVIEW

**Comment by droazen** [comment]
@lbergelson Could you review the gradle build script in this PR?

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// See the second 'repositories' section below for the actual dependencies of GATK itself
+buildscript {
+    repositories {
+        mavenCentral()
+     }
+}
+
+plugins {
+    id "java"
+    id "application"
```
Comment (lbergelson) [inline build.gradle:11]: Don't need the application plugin since it's not a runnable thing.

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+//Delete the windows script - we never test on Windows so let's not pretend it works
+startScripts {
```
Comment (lbergelson) [inline build.gradle:15]: same with the start scripts, not needed

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+// This is a hack to disable the java 8 default javadoc lint until we fix the html formatting
+if (JavaVersion.current().isJava8Compatible()) {
+    tasks.withType(Javadoc) {
+        options.addStringOption('Xdoclint:none', '-quiet')
+    }
+}
+
+task installAll{  dependsOn installDist }
```
Comment (lbergelson) [inline build.gradle:85]: don't need this and the following lines

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+//Delete the windows script - we never test on Windows so let's not pretend it works
+startScripts {
+    doLast {
+        delete windowsScript
+    }
+}
+
+repositories {
```
Comment (lbergelson) [inline build.gradle:21]: this has no dependencies so it doesn't need any repositories here

File: build.gradle [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+compileTestJava {
+  options.compilerArgs = ['-proc:none', '-Xlint:all','-Werror','-Xdiags:verbose']
+}
+
+
+sourceCompatibility = 1.8
+targetCompatibility = 1.8
+
+def String deriveVersion(){
```
Comment (lbergelson) [inline build.gradle:42]: I'm going to simplify how we do version handling with a plugin I found, but that can happen later.

File: src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package org.broadinstitute.gatk.nativebindings;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java:1]: Should this just be org.broadinstitute.nativebinding ?  It's not really gatk specific.

File: build.gradle [hunk 7]
```diff
@@ -0,0 +1,89 @@
+//Note: this section 'buildscript` is only for the dependencies of the buildscript itself.
+// See the second 'repositories' section below for the actual dependencies of GATK itself
```
Comment (lbergelson) [inline build.gradle:2]: This line talking about the gatk should be updated since it's a different repo.

File: build.gradle [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+repositories {
+    mavenCentral()
+    jcenter()
+
+    maven {
+        url "https://artifactory.broadinstitute.org/artifactory/libs-snapshot/" //for htsjdk snapshots
+    }
+}
+
+//NOTE: we ignore contracts for now
```
Comment (lbergelson) [inline build.gradle:30]: drop this contracts line

**Comment by lbergelson** [comment]
@droazen Build script looks fine but most of it can be deleted.  I might drop all the automatic version determination and just set the version to 1.0.0 manually.  We'll probably want to add the maven and signing plugins to help with releasing it to central.  We could just upload it manually, but getting the signing stuff right is going to be a pain that way.

File: src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java (hunk 6 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java:1]: also, gatk itself is org.broadinstitute.hellbender, not org.broadinstitute.gatk

**Comment by akiezun** [comment]
can we merge this and then @lbergelson will improve the build script? that'll be much more efficient i think

**Comment by lbergelson** [comment]
@akiezun fine with me

File: src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java (hunk 6 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java:1]: I think `org.broadinstitute.gatk.nativebindings` is fine given that the repo is "gatk-native-bindings" (ie., native bindings for the GATK specifically), and given that the hellbender package will eventually be renamed to gatk.

File: src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import java.nio.file.Path;
+
+/**
+ * Common interface for all native libraries.
+ */
+public interface NativeLibrary {
+    /**
+     * Returns the path to this native library.
+     * Typically this is the path to the shared object file. GATK will use this path to load the native library
+     * using <code>System.load(path.toString())</code>
```
Comment (droazen) [inline src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java]: Path will likely be location on the classpath, in which case `System.load()` will not be run on this path directly -- the library will have to be extracted to a temp location first before loading. Update docs to clarify.

**Comment by droazen** [comment]
Merge after fixing inaccurate docs for the `NativeLibrary.getLibraryPath()` method. Back to @akiezun

File: src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java:1]: excactly

File: src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java (hunk 9 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/gatk/nativebindings/NativeLibrary.java]: I'm just going to remove this part of the doc for now.

**Pull request merged**